server/src/com/cloud/resource/ResourceManagerImpl.java

```diff
@@ -1296,7 +1296,8 @@ public boolean checkAndMaintain(final long hostId) {
             if (host.getType() != Host.Type.Storage) {
                 final List<VMInstanceVO> vos = _vmDao.listByHostId(hostId);
                 final List<VMInstanceVO> vosMigrating = _vmDao.listVmsMigratingFromHost(hostId);
-                if (vos.isEmpty() && vosMigrating.isEmpty()) {
+                final List<VMInstanceVO> failedMigratedVms = _vmDao.listRunningVmsByHostAndLastHostSameId(hostId);
+                if (vos.isEmpty() && vosMigrating.isEmpty() && failedMigratedVms.isEmpty()) {
```

Comment:
So this condition means
private boolean allMigrationWorkDone(){
    final List<VMInstanceVO> vos = _vmDao.listByHostId(hostId);
    final List<VMInstanceVO> vosMigrating = _vmDao.listVmsMigratingFromHost(hostId);
    final List<VMInstanceVO> failedMigratedVms = _vmDao.listRunningVmsByHostAndLastHostSameId(hostId);

    return vos.isEmpty() && vosMigrating.isEmpty() && failedMigratedVms.isEmpty();
}
let's create that method

Reply:
This also allows at least some unit testing....